core/src/main/java/org/mule/runtime/core/internal/construct/BackPressureStrategySelector.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.construct;
+
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.api.source.MessageSource;
+import org.mule.runtime.core.internal.util.MessagingExceptionResolver;
+
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * Implements the different backpressure handling strategies, and checks against a
+ * {@link org.mule.runtime.core.api.processor.strategy.ProcessingStrategy} whether or not backpressure is fired, before and event
+ * being processing.
+ */
+public class BackPressureStrategySelector {
+
+  private static int EVENT_LOOP_SCHEDULER_BUSY_RETRY_INTERVAL_MS = 2;
+
+  private AbstractPipeline abstractPipeline;
+  private final MessagingExceptionResolver exceptionResolver;
+
+  public BackPressureStrategySelector(AbstractPipeline abstractPipeline) {
+    this.abstractPipeline = abstractPipeline;
+    exceptionResolver = new MessagingExceptionResolver(abstractPipeline);
+  }
+
+  /**
+   * Wait backpressure strategy. Implements a busy-wait strategy.
+   *
+   * @param event the event about to begin processing
+   * @throws FlowBackPressureException
+   */
+  protected void checkWithWaitStrategy(CoreEvent event)
+      throws FlowBackPressureException {
+    boolean accepted = false;
+    while (!accepted) {
+      try {
+        abstractPipeline.getProcessingStrategy().checkBackpressureAccepting(event);
+        accepted = true;
+      } catch (RejectedExecutionException ree) {
+        // TODO MULE-16106 Add a callback for WAIT back pressure applied on the source
+        try {
+          Thread.sleep(EVENT_LOOP_SCHEDULER_BUSY_RETRY_INTERVAL_MS);
+        } catch (InterruptedException e) {
+          FlowBackPressureException exception = new FlowBackPressureException(abstractPipeline.getName(), ree);
```

Comment:
reset interrupted flag

Reply:
Using Thread.interrupted() method. Is this ok?